functions_advanced.ipynb

```diff
@@ -0,0 +1,4916 @@
+   "id": "62bd58ff-2b79-4f3f-ad6f-06d22736b53c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "next(sq)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0f2ce2d7-2ec6-428f-aa8c-94cc6d6a65b2",
+   "metadata": {},
+   "source": [
+    "But since we now have consumed **2 elements** from `sq`, when we use `enumerate` it will also have two less items from `sq`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9070f983-4938-4ee1-a97a-b85ecfc2c5bd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "next(enum_sq)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "040e46b2-206a-4051-b548-62acec5287be",
+   "metadata": {},
+   "source": [
+    "And this might not be what you expected: the value is the **third** element of `sq` ($2^2$), while the index is `0`, as if we were starting from the beginning.\n",
+    "From the point of view of the generator returned by `enumerate`, **we are at the beginning**.\n",
+    "\n",
+    "So, beware when you are combining multiple generators, and think carefully what's the behavior you expect."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f5214c67-0f92-4a7f-a768-eee90dd14bbb",
+   "metadata": {},
+   "source": [
+    "## Exercises"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "69a49f30-1d16-4c84-89bc-30d6bd7421f2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%reload_ext tutorial.tests.testsuite"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "777251e1-d655-4e36-a41c-ddefe020d1b7",
+   "metadata": {
+    "jp-MarkdownHeadingCollapsed": true
+   },
+   "source": [
+    "### Password checker factory"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bbec8f73-9b4f-49bc-bd59-c3db2ed6132d",
+   "metadata": {
+    "jp-MarkdownHeadingCollapsed": true
+   },
+   "source": [
+    "Create a function called `password_checker_factory` that can be used to generate different password checkers.\n",
+    "This function will take **four parameters**: `min_uppercase`, `min_lowercase`, `min_punctuation`, and `min_digits`.\n",
+    "They represents the constraints on a given password:\n",
+    "\n",
+    "1. The minimum number of uppercase letters.\n",
+    "2. The minimum number of lowercase letters.\n",
+    "3. The minimum number of punctuation characters.\n",
+    "4. The minimum number of digits.\n",
+    "\n",
+    "\n",
+    "The create_password_checker function generates another function that assesses a given password (string).\n",
+    "This resulting function returns a **tuple with two elements**:\n",
+    "\n",
+    "1. The first element is a **boolean** indicating if the password passed validation.\n",
+    "2. The second element is a **dictionary** showing how much the password exceeds the minimum requirements for \"uppercase,\" \"lowercase,\" \"punctuation,\" and \"digits.\" Negative values indicate that the minimum requirements are **not** met."
```

Comment:
```suggestion
    "2. The second element is a **dictionary** showing by how many digits the password exceeds the minimum requirements for \"uppercase,\" \"lowercase,\" \"punctuation,\" and \"digits.\" Negative values indicate that the minimum requirements are **not** met."
```

Reply:
"Digits" is misleading here. Need to think of a better reformulation.

Reply:
Reformulated ✅